### PR TITLE
Feature - @InterfaceType decorator

### PR DESCRIPTION
**Pull request by felipesabino**
- Added `@InterfaceType` decorator - #28 
- Inferring `@ObjectType` classes that implements the interface decorator but are not directly used anywhere else.
- Added `@InterfaceType` to #23 issue description for future improvement
- Added `@InterfaceType` to README

### CODE REVIEW

File: src/array.utils.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Concats two arrays
+ * @param itemsA fist array argument
```
Comment (marcelorisse) [inline src/array.utils.ts]: Typo

File: src/array.utils.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * Flattens an array with nested arrays
+ * @param collection the array argument
+ */
+export const flatten = (collection: any[]) => flatMap(items => {
+  if (items.constructor === Array) {
+    return flatMap(item => item, items);
+  } else {
+    return items;
+  }
+}, collection);
```
Comment (thiago-soliveira) [inline src/array.utils.ts:25]: What about `lodash.flatten` and `lodash.concat`?

**Review by thiago-soliveira — APPROVED** [review]
Very nice!

File: src/array.utils.ts (hunk 2 above)
Comment (felipesabino) [reply src/array.utils.ts:25]: @thiago-soliveira Thought about that but I prefer avoid adding extra dependencies for very simple use cases whenever possible

File: src/type-factory/interface.type-factory.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import { getMetadataBuilder } from '../metadata-builder';
+import { objectTypeFactory } from './object.type-factory';
+import { fieldTypeFactory } from './field.type-factory';
+
+let interfaceTypeCache: { [key: string]: any } = {};
+
+export function interfaceTypeFactory(target: any): graphql.GraphQLInterfaceType | undefined {
+  return getMetadataBuilder().buildInterfaceTypeMetadata(target)
+    .map(metadata => {
+      if (interfaceTypeCache[metadata.name]) {
```
Comment (tibawatanabe) [inline src/type-factory/interface.type-factory.ts]: Hum... what about:

```ts
if (!interfaceTypeCache[metadata.name]) {
  interfaceTypeCache[metadata.name] = new ...
}

return interfaceTypeCache[metadata.name];
```

File: src/type-factory/object.type-factory.ts [hunk 4]
```diff
@@ -26,26 +28,44 @@ export function objectTypeFactory(target: Function, isInput: boolean = false): g
               // TODO write test
               throw new SchemaFactoryError('', SchemaFactoryErrorType.INVALID_OBJECT_TYPE_METADATA);
           }
-          const fieldMetadataList: FieldMetadata[] = getMetadataBuilder().buildFieldMetadata(target.prototype);
+          const fieldMetadataList: { target: Function, field: FieldMetadata}[] = getMetadataBuilder()
+            .buildFieldMetadata(target.prototype)
+            .map(field => ({
+              target: target,
```
Comment (tibawatanabe) [inline src/type-factory/object.type-factory.ts]: possible shorthand:

```ts
({ target, field })
```

File: src/metadata-builder/metadata.builder.ts [hunk 5]
```diff
@@ -49,6 +61,7 @@ export class MetadataBuilder {
         name: arg.name,
         description: arg.description,
         isInput: arg.isInput,
+        interfaces: Array.prototype.concat.apply([], arg.interfaces.map(this.buildInterfaceTypeMetadata)),
```
Comment (tibawatanabe) [inline src/metadata-builder/metadata.builder.ts]: We can use flatmap here

**Review by tibawatanabe — APPROVED** [review]

**Pull request merged**